File: packages/telemetry/python/examples/test_bedrock.py
"""
AWS Bedrock — Latitude telemetry example.

Uses the Bedrock `converse` / `converse_stream` API (better instrumentation
support than `invoke_model`). Credentials are resolved via the default AWS
credential chain (env vars, shared config, SSO, instance role).

Required env vars:
- LATITUDE_API_KEY
- LATITUDE_PROJECT_SLUG
- AWS credentials via the default chain (and AWS_REGION; default eu-central-1)

Install: uv add boto3
"""

import os
import uuid

import boto3

from latitude_telemetry import Latitude, capture

latitude = Latitude(
    api_key=os.environ["LATITUDE_API_KEY"],
    project=os.environ["LATITUDE_PROJECT_SLUG"],
    instrumentations={"bedrock": boto3},
    disable_batch=True,
)

PROVIDER = "bedrock"
# Cross-region inference profile for eu-central-1; the instrumentor strips the
# `eu.` prefix -> vendor "anthropic", model "claude-opus-4-8" for cost lookup.
MODEL = "eu.anthropic.claude-opus-4-8"
MAX_TOKENS = 1024
REGION = os.environ.get("AWS_REGION", "eu-central-1")
# converse `system` is the out-of-band system field — verify it lands in systemInstructions.
SYSTEM = "You are a helpful assistant participating in a telemetry QA test. Keep answers concise."
SESSION_ID = f"{PROVIDER}-{uuid.uuid4().hex[:8]}"


def _client():
    return boto3.client("bedrock-runtime", region_name=REGION)


def _ctx(scenario: str, *extra_tags: str) -> dict:
    return {
        "tags": ["example", PROVIDER, "bedrock-py", *extra_tags],
        "session_id": SESSION_ID,
        "user_id": "example-user",
        "metadata": {"scenario": scenario, "environment": "local"},
    }


def chat() -> str:
    response = _client().converse(
        modelId=MODEL,
        system=[{"text": SYSTEM}],
        messages=[{"role": "user", "content": [{"text": "Say 'Hello from Bedrock!' in exactly 5 words."}]}],
        inferenceConfig={"maxTokens": MAX_TOKENS},
    )
    return response["output"]["message"]["content"][0]["text"]


def stream() -> str:
    response = _client().converse_stream(
        modelId=MODEL,
        system=[{"text": SYSTEM}],
        messages=[{"role": "user", "content": [{"text": "Say 'Hello from Bedrock stream!' in exactly 6 words."}]}],
        inferenceConfig={"maxTokens": MAX_TOKENS},
    )
    chunks: list[str] = []
    for event in response["stream"]:
        delta = event.get("contentBlockDelta", {}).get("delta", {})
        if "text" in delta:
            chunks.append(delta["text"])
    return "".join(chunks)


def tool_conversation() -> str:
    client = _client()
    tool_config = {
        "tools": [
            {
                "toolSpec": {
                    "name": "get_weather",
                    "description": "Get the current weather for a city",
                    "inputSchema": {
                        "json": {
                            "type": "object",
                            "properties": {"city": {"type": "string"}},
                            "required": ["city"],
                        }
                    },
                }
            }
        ]
    }
    messages = [
        {
            "role": "user",
            "content": [
                {"text": "What's the weather in San Francisco? Use get_weather, then answer in one short sentence."}
            ],
        }
    ]

    first = client.converse(
        modelId=MODEL,
        system=[{"text": SYSTEM}],
        messages=messages,
        toolConfig=tool_config,
        inferenceConfig={"maxTokens": MAX_TOKENS},
    )
    out_message = first["output"]["message"]
    messages.append(out_message)
    tool_use = next(b["toolUse"] for b in out_message["content"] if "toolUse" in b)
    messages.append(
        {
            "role": "user",
            "content": [
                {
                    "toolResult": {
                        "toolUseId": tool_use["toolUseId"],
                        "content": [{"json": {"city": "San Francisco", "temperatureC": 21, "conditions": "sunny"}}],
                    }
                }
            ],
        }
    )

    second = client.converse(
        modelId=MODEL,
        system=[{"text": SYSTEM}],
        messages=messages,
        toolConfig=tool_config,
        inferenceConfig={"maxTokens": MAX_TOKENS},
    )
    return "".join(b.get("text", "") for b in second["output"]["message"]["content"])


if __name__ == "__main__":
    capture("bedrock-chat-capture", chat, _ctx("chat"))
    capture("bedrock-stream-capture", stream, _ctx("stream", "stream"))
    capture("bedrock-tools-capture", tool_conversation, _ctx("tools", "tools"))

    latitude.flush()

File: packages/telemetry/typescript/examples/test_bedrock.ts
/**
 * AWS Bedrock — Latitude telemetry example.
 *
 * Required env vars:
 * - LATITUDE_API_KEY
 * - LATITUDE_PROJECT_SLUG
 * - AWS credentials via the default chain (env vars, shared config, SSO, instance role…)
 * - AWS_REGION (default: eu-central-1)
 *
 * Install: npm install @aws-sdk/client-bedrock-runtime
 */

import { randomUUID } from "node:crypto"
import * as BedrockSDK from "@aws-sdk/client-bedrock-runtime"
import {
  BedrockRuntimeClient,
  InvokeModelCommand,
  InvokeModelWithResponseStreamCommand,
} from "@aws-sdk/client-bedrock-runtime"
import { capture, Latitude } from "../src"

const latitude = new Latitude({
  apiKey: process.env.LATITUDE_API_KEY!,
  project: process.env.LATITUDE_PROJECT_SLUG!,
  disableBatch: true,
  instrumentations: { bedrock: BedrockSDK },
})

const PROVIDER = "bedrock"
// Cross-region inference profile id; the instrumentation strips the `eu.` prefix to detect the vendor.
const MODEL = "eu.anthropic.claude-opus-4-8"
const REGION = process.env.AWS_REGION || "eu-central-1"
const ANTHROPIC_VERSION = "bedrock-2023-05-31"
const MAX_TOKENS = 1024
const SYSTEM = "You are a helpful assistant participating in a telemetry QA test. Keep answers concise."
const SESSION_ID = `${PROVIDER}-${randomUUID().slice(0, 8)}`

type ContentBlock = { type: string; [key: string]: unknown }
type Message = { role: "user" | "assistant"; content: string | ContentBlock[] }

function ctx(scenario: string, ...extraTags: string[]) {
  return {
    tags: ["example", PROVIDER, "bedrock-ts", ...extraTags],
    sessionId: SESSION_ID,
    userId: "example-user",
    metadata: { scenario, environment: "local" },
  }
}

function client() {
  return new BedrockRuntimeClient({ region: REGION })
}

async function invoke(body: Record<string, unknown>) {
  const response = await client().send(
    new InvokeModelCommand({
      modelId: MODEL,
      contentType: "application/json",
      accept: "application/json",
      body: JSON.stringify(body),
    }),
  )
  return JSON.parse(new TextDecoder().decode(response.body)) as { content: ContentBlock[] }
}

async function chat() {
  const res = await invoke({
    anthropic_version: ANTHROPIC_VERSION,
    max_tokens: MAX_TOKENS,
    system: SYSTEM,
    messages: [{ role: "user", content: "Say 'Hello from Bedrock!' in exactly 5 words." }],
  })
  return res.content.find((b) => b.type === "text")?.text
}

async function stream() {
  const response = await client().send(
    new InvokeModelWithResponseStreamCommand({
      modelId: MODEL,
      contentType: "application/json",
      accept: "application/json",
      body: JSON.stringify({
        anthropic_version: ANTHROPIC_VERSION,
        max_tokens: MAX_TOKENS,
        system: SYSTEM,
        messages: [{ role: "user", content: "Say 'Hello from Bedrock stream!' in exactly 6 words." }],
      }),
    }),
  )

  const chunks: string[] = []
  for await (const event of response.body ?? []) {
    const bytes = event.chunk?.bytes
    if (!bytes) continue
    const evt = JSON.parse(new TextDecoder().decode(bytes))
    if (evt.type === "content_block_delta" && evt.delta?.type === "text_delta") {
      chunks.push(evt.delta.text)
    }
  }
  return chunks.join("")
}

async function toolConversation() {
  const tools = [
    {
      name: "get_weather",
      description: "Get the current weather for a city",
      input_schema: {
        type: "object",
        properties: { city: { type: "string" } },
        required: ["city"],
      },
    },
  ]

  const messages: Message[] = [
    {
      role: "user",
      content: "What's the weather in San Francisco? Use get_weather, then answer in one short sentence.",
    },
  ]

  const first = await invoke({ anthropic_version: ANTHROPIC_VERSION, max_tokens: MAX_TOKENS, system: SYSTEM, tools, messages })
  const toolUse = first.content.find((b) => b.type === "tool_use")
  messages.push({ role: "assistant", content: first.content })
  messages.push({
    role: "user",
    content: [
      {
        type: "tool_result",
        tool_use_id: toolUse!.id as string,
        content: JSON.stringify({ city: "San Francisco", temperatureC: 21, conditions: "sunny" }),
      },
    ],
  })

  const second = await invoke({ anthropic_version: ANTHROPIC_VERSION, max_tokens: MAX_TOKENS, system: SYSTEM, tools, messages })
  return second.content.find((b) => b.type === "text")?.text
}

async function main() {
  await latitude.ready

  await capture("bedrock-chat-capture", chat, ctx("chat"))
  await capture("bedrock-stream-capture", stream, ctx("stream", "stream"))
  await capture("bedrock-tools-capture", toolConversation, ctx("tools", "tools"))

  await latitude.flush()
  await latitude.shutdown()
}

main().catch(console.error)
